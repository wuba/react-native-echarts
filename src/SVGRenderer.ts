import env from 'zrender/lib/core/env';
import SVGPainter from 'zrender/lib/svg/Painter';
import { vNodeToString } from './SVGCore';
import { updateAttrs } from 'zrender/lib/svg/patch';
import type Storage from 'zrender/lib/Storage';
import { DOMParser } from '@xmldom/xmldom';

env.svgSupported = true;

interface SVGPainterOption {
  width?: number;
  height?: number;
  ssr?: boolean;
}

interface RootProps extends HTMLElement {
  getChartSize: () => {
    width: number;
    height: number;
  };
}

global.DOMParser = DOMParser;
class CustomSVGPainter extends SVGPainter {
  constructor(
    root: RootProps,
    storage: Storage,
    opts: SVGPainterOption,
    id: number
  ) {
    // Prioritize taking the width and height set in the configuration;
    // if not available, then take the width and height set in the styles.
    const { width, height } = root.getChartSize();
    opts.width = opts.width || width;
    opts.height = opts.height || height;
    // @ts-ignore
    super(null, storage, opts);
    // @ts-ignore
    this._svgDom = this._oldVNode.elm = root.elm;
    // @ts-ignore
    this._svgDom.setZrenderId?.(id);
    // @ts-ignore
    updateAttrs(null, this._oldVNode);
    this.root = root;
  }
  refresh() {
    const vnode = this.renderToVNode({
      willUpdate: true,
    });
    // Disable user selection.
    vnode.attrs.style = 'position:absolute;left:0;top:0;user-select:none';
    // @ts-ignore
    if (this._svgDom.patch) {
      // @ts-ignore
      this._svgDom.patch(this._oldVNode, vnode);
      // @ts-ignore
    } else if (this._svgDom.patchString) {
      // @ts-ignore
      this._svgDom.patchString(this._oldVNode, vNodeToString(vnode));
    }
    // @ts-ignore
    this._oldVNode = vnode;
  }
  toDataURL(base64?: boolean): string {
    // @ts-ignore
    return this._svgDom.makeImageSnapshot?.() || super.toDataURL(base64);
  }
}

export function SVGRenderer(registers: any) {
  registers.registerPainter('svg', CustomSVGPainter);
}
