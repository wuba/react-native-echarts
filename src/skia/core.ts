export type CSSSelectorVNode = Record<string, string>
export type CSSAnimationVNode = Record<string, Record<string, string>>
export type SVGVNodeAttrs = Record<string, string | number | undefined | boolean>
export interface SVGVNode {
  tag: string,
  attrs: SVGVNodeAttrs,
  children?: SVGVNode[],
  text?: string

  // For patching
  elm?: Node
  key: string
};
export interface BrushScope {
  zrId: string

  shadowCache: Record<string, string>
  gradientCache: Record<string, string>
  patternCache: Record<string, string>
  clipPathCache: Record<string, string>

  defs: Record<string, SVGVNode>

  cssNodes: Record<string, CSSSelectorVNode>
  cssAnims: Record<string, Record<string, Record<string, string>>>
  /**
   * Cache for css style string, mapping from style string to class name.
   */
  cssStyleCache: Record<string, string>

  cssAnimIdx: number

  shadowIdx: number
  gradientIdx: number
  patternIdx: number
  clipPathIdx: number
  // configs
  /**
   * If create animates nodes.
   */
  animation?: boolean,
  /**
   * If create emphasis styles.
   */
  emphasis?: boolean,

  /**
   * If will update. Some optimization for string generation can't be applied.
   */
  willUpdate?: boolean

  /**
   * If compress the output string.
   */
  compress?: boolean

  ssr?: boolean
}

export function createBrushScope(zrId: string): BrushScope {
  return {
    zrId,
    shadowCache: {},
    patternCache: {},
    gradientCache: {},
    clipPathCache: {},
    defs: {},

    cssNodes: {},
    cssAnims: {},
    cssStyleCache: {},

    cssAnimIdx: 0,

    shadowIdx: 0,
    gradientIdx: 0,
    patternIdx: 0,
    clipPathIdx: 0
  };
}