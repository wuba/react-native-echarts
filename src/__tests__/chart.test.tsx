/**
 * @jest-environment @shopify/react-native-skia/jestEnv.mjs
 */
import React, { useEffect, useRef } from 'react';
import { Dimensions } from 'react-native';
import { render } from '@testing-library/react-native';
import SkiaComponent from '../skiaChart';
import SVGComponent from '../svgChart';
import { SVGRenderer } from '../SVGRenderer';
import * as echarts from 'echarts/core';
import { BarChart } from 'echarts/charts';
import {
  TitleComponent,
  TooltipComponent,
  GridComponent,
} from 'echarts/components';
echarts.use([
  TitleComponent,
  TooltipComponent,
  GridComponent,
  SVGRenderer,
  BarChart,
]);
const Components = [SkiaComponent, SVGComponent];
const E_HEIGHT = 250;
const E_WIDTH = Dimensions.get('screen').width;//750
const option = {
  xAxis: {
    type: 'category',
    data: ['Mon', 'Tue'],
  },
  yAxis: {
    type: 'value',
  },
  series: [
    {
      data: [120, 200],
      type: 'bar',
    },
  ],
};
function Chart({ option, Component }: any) {
  const ref = useRef<any>(null);
  useEffect(() => {
    let chart: any;
    if (ref.current) {
      // @ts-ignore
      chart = echarts.init(ref.current, 'light', {
        renderer: 'svg',
        width: E_WIDTH,
        height: E_HEIGHT,
      });
      chart.setOption(option);
    }
    return () => chart?.dispose();
  }, [option]);

  return <Component ref={ref} />;
}
Components.forEach((Component) => {
  describe(`${Component.displayName} Chart` || 'unknown', () => {
    it('renders correctly', () => {
      const { toJSON } = render(
        <Chart option={option} Component={Component} />
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
