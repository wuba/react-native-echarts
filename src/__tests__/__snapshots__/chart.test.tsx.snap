// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkiaChart Chart renders correctly 1`] = `
<View
  style={
    {
      "height": 250,
      "width": 750,
    }
  }
  testID="component"
>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      {
        "backgroundColor": "rgba(0,0,0,0)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
    testID="pan-responder-handler"
  />
</View>
`;

exports[`SvgChart Chart renders correctly 1`] = `
<View
  style={
    {
      "height": 250,
      "width": 750,
    }
  }
  testID="component"
>
  <RNSVGSvgView
    bbHeight="250"
    bbWidth="750"
    focusable={false}
    height={250}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 250,
          "width": 750,
        },
      ]
    }
    width={750}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGRect
        fill={null}
        height="250"
        propList={
          [
            "fill",
          ]
        }
        width="750"
        x="0"
        y="0"
      />
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGPath
          d="M75 180.5L675 180.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4292929265,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M75 150.5L675 150.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4292929265,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M75 120.5L675 120.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4292929265,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M75 90.5L675 90.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4292929265,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M75 60.5L675 60.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4292929265,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M75 180.5L675 180.5"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
              "strokeLinecap",
            ]
          }
          stroke={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          strokeLinecap={1}
        />
        <RNSVGPath
          d="M75.5 180L75.5 185"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M375.5 180L375.5 185"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
        />
        <RNSVGPath
          d="M675.5 180L675.5 185"
          fill={
            {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
              "stroke",
            ]
          }
          stroke={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
        />
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "end",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              67,
              180,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={[]}
        >
          <RNSVGTSpan
            content="0"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "end",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              67,
              150,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={[]}
        >
          <RNSVGTSpan
            content="50"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "end",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              67,
              120,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={[]}
        >
          <RNSVGTSpan
            content="100"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "end",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              67,
              90,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={[]}
        >
          <RNSVGTSpan
            content="150"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "end",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              67,
              60,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={[]}
        >
          <RNSVGTSpan
            content="200"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "middle",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              225,
              188,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={
            [
              "6",
            ]
          }
        >
          <RNSVGTSpan
            content="Mon"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGText
          alignmentBaseline="middle"
          content={null}
          dx={[]}
          dy={[]}
          fill={
            {
              "payload": 4285427833,
              "type": 0,
            }
          }
          fillOpacity={1}
          font={
            {
              "fontFamily": "Helvetica Neue",
              "fontSize": "12px",
              "textAnchor": "middle",
            }
          }
          matrix={
            [
              1,
              0,
              0,
              1,
              525,
              188,
            ]
          }
          propList={
            [
              "fill",
              "fillOpacity",
            ]
          }
          rotate={[]}
          x={[]}
          y={
            [
              "6",
            ]
          }
        >
          <RNSVGTSpan
            content="Tue"
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={{}}
            rotate={[]}
            x={[]}
            y={[]}
          />
        </RNSVGText>
        <RNSVGPath
          d="M121.5 180l207 0l0 -72l-207 0Z"
          fill={
            {
              "payload": 4281836250,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
        <RNSVGPath
          d="M421.5 180l207 0l0 -120l-207 0Z"
          fill={
            {
              "payload": 4281836250,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      {
        "backgroundColor": "rgba(0,0,0,0)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
    testID="pan-responder-handler"
  />
</View>
`;
